### ISSUE TITLE
Poor default autocompletion heuristics

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#15013]
## Description

The default heuristics for autocompletion introduced in JupyterLab 4.0 with https://github.com/jupyterlab/jupyterlab/pull/11969 are not very good.
- If I type a first character nothing happens. It only triggers on the second character.
- It triggers on space (to reproduce create a file/folder with space in name and use Ipython)
- Then it also triggers after closing parenthesis so when I press enter to go to next line it instead completes the text.

![Screenshot from 2023-08-24 21-51-17](https://github.com/jupyterlab/jupyterlab/assets/5832902/2870bec8-1d1d-4c2f-86c9-d4d598239d56)

It seems to also use some logic conditioning on a dot (`.`) but this is very much not cross-language compatible.

## Reproduce

![poor-defaults](https://github.com/jupyterlab/jupyterlab/assets/5832902/e6af4a4e-55ad-43da-b216-98c06953b8fe)

## Expected behavior

In LSP extension we were using CodeMirror 5 token type to decide if the completer should be triggered. Maybe there is a way to make use of CM6 language data somehow but there does not seem to be a dedicated method for it.

For the newline issue we could make it easier to remap "accept completion" to "tab" rather than current "enter". Currently it involves manually changing three shortucuts (for notebook, file editor and console).

## Context

- Browser and version: Chrome 115
- JupyterLab version: 4.0.5

**Comment by trungleduc** [comment]
> If I type a first character nothing happens. It only triggers on the second character.

This is due to these 2 callbacks https://github.com/jupyterlab/jupyterlab/blob/7b1effc02bb11e237f98bdf3c888862bc183e15d/packages/completer/src/handler.ts#L224-L230 

and 

https://github.com/jupyterlab/jupyterlab/blob/7b1effc02bb11e237f98bdf3c888862bc183e15d/packages/completer/src/handler.ts#L283-L290

Both are called when users type the first letter but the the callback used to show the completer is called before the one used to enable the completer. I think this bug exists even before 4.0

> It triggers on space (to reproduce create a file/folder with space in name and use Ipython)

Since IPython proposes auto-completion of the file path which can be anything, triggering and showing the results containing space is logical.

>In LSP extension we were using CodeMirror 5 token type to decide if the completer should be triggered. Maybe there is a way to make use of CM6 language data somehow but there does not seem to be a dedicated method for it.

Same here, since users can have auto-completion results from the kernel for the file name/path or Ipython magics, limiting the token type might eliminate these features.

**Comment by krassowski** [comment]
No, this is not what I am referring too. I am referring specifically to the logic in:

https://github.com/jupyterlab/jupyterlab/blob/7b1effc02bb11e237f98bdf3c888862bc183e15d/packages/completer/src/default/kernelprovider.ts#L123-L142

There was no autocompletion of this kind in JupterLab prior to 4.0. What I am saying is that it is very bad UX.

> Since IPython proposes auto-completion of the file path which can be anything, triggering and showing the results containing space is logical.

No, because completer should not be invoked when pressing space. Not by default. This is not what I as a user expect.

> Same here, since users can have auto-completion results from the kernel for the file name/path or Ipython magics, limiting the token type might eliminate these features

No, it would not eliminate any feature. I am only saying that by default space and closing parenthesis, colon, etc should not trigger autocompletion on key press. Imaging that you type `def func():` and press enter to go to next line but instead it completes the editor with `my file: name`. This is incredibly annoying, especially because undo will result in removing both fragments (because the undo history gets merged for quickly typed phrases since `y` migration)
- users will be able to get these completions by pressing <kbd>tab</kbd> anyways
- users will be able to go to settings and change what kind of suppression is active

For reference this is what was offered by the LSP extension:
![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/9ae3a8f4-313c-4e22-ae6e-9bd769a3dfd6)

**Comment by krassowski** [comment]
Does not even need a file with colon in name:

![example](https://github.com/jupyterlab/jupyterlab/assets/5832902/e7a7e20e-62a5-44a2-9a13-416f5a68e227)

**Comment by trungleduc** [comment]
> No, this is not what I am referring too. I am referring specifically to the logic in:
> 
> https://github.com/jupyterlab/jupyterlab/blob/7b1effc02bb11e237f98bdf3c888862bc183e15d/packages/completer/src/default/kernelprovider.ts#L123-L142
> 
> There was no autocompletion of this kind in JupterLab prior to 4.0. What I am saying is that it is very bad UX.
> 

This `shouldShowContinuousHint` function is called from the `onTextChanged` callback, which is called before the `onSelectionsChanged` callback. On the first letter, the completer is disabled and will be enabled by the logic inside `onSelectionsChanged`, but since the `onTextChanged` is called first, the completer is still disabled and it will return early. But I agree that the logic of `shouldShowContinuousHint` needs to be improved to be kernel-agnostic.


> > Since IPython proposes auto-completion of the file path which can be anything, triggering and showing the results containing space is logical.
> 
> No, because completer should not be invoked when pressing space. Not by default. This is not what I as a user expect.
> 
> > Same here, since users can have auto-completion results from the kernel for the file name/path or Ipython magics, limiting the token type might eliminate these features
> 
> No, it would not eliminate any feature. I am only saying that by default space and closing parenthesis, colon, etc should not trigger autocompletion on key press. Imaging that you type `def func():` and press enter to go to next line but instead it completes the editor with `my file: name`. This is incredibly annoying, especially because undo will result in removing both fragments (because the undo history gets merged for quickly typed phrases since `y` migration)
> 
> * users will be able to get these completions by pressing tab anyways
> * users will be able to go to settings and change what kind of suppression is active
> 


I agree, that the autocompletion mode needs to be refined, the current implementation is just a shortcut of the manual mode.

**Comment by krassowski** [comment]
> On the first letter, the completer is disabled and will be enabled by the logic inside `onSelectionsChanged`, but since the `onTextChanged` is called first, the completer is still disabled and it will return early.

Agreed, this logic is due a refactor, ideally in a way enabling https://github.com/jupyterlab/jupyterlab/issues/13006.

**Comment by Xonxt** [comment]
I'm honestly appalled this is still a thing. This autocomplete is TOO eager. Any updates on mitigating this behaviour?

![jupyterlab](https://github.com/jupyterlab/jupyterlab/assets/5681661/27f2439b-7253-44ba-9f19-7753a0c2db1f)

**Comment by krassowski** [comment]
Yes, I also saw many reports of frustrated users.

> Any updates on mitigating this behaviour?

you can install [jupyterlab-lsp](https://github.com/jupyter-lsp/jupyterlab-lsp) and enable its continuous hinting which has sane heuristics.

If anyone is interested in supporting this effort I am happy to work on a good default implementation of autocompletion but until the end of the year my spare time for volunteering is thin.

**Comment by adriengcql** [comment]
> you can install [jupyterlab-lsp](https://github.com/jupyter-lsp/jupyterlab-lsp) and enable its continuous hinting which has sane heuristics.

Do you know a way to have the filepath completion with jupyterlab-lsp ? Otherwise we are missing it when disabling the _kernel_ completion provider.

Another workaround could be to enable continuous hinting only with the lsp provider and keep the other providers when hitting _tab_, but I couldn't find any configuration to achieve that.

**Comment by Yiannis128** [comment]
I've disabled autocomplete, it gets too much in the way... It would be such a nice feature such a shame :/

**Comment by transistorgrab** [comment]
Found this today, am annoyed by the same behaviour.
Constantly need to hit Esc to close the popup.
Still often accidentally get some random crap inserted when just hitting Enter to start a new line…

**Comment by YongcaiHuang** [comment]
To be honest, I also agree and think the autocompletion is too aggressive to code smoothly. The issue has been open since 2023 with milestone 4.0.x. Yet, we are heading to 4.6, and the problem remains. I believe solving this could largely enhance the coding experience.
